firebase-firestore/src/main/java/com/google/firebase/firestore/remote/Datastore.java

```diff
@@ -222,19 +222,41 @@ public void onClose(Status status) {
     return completionSource.getTask();
   }
 
-  public Task<Long> runCountQuery(Query query) {
+  public Task<Map<String, Value>> runAggregateQuery(
+      Query query, List<AggregateField> aggregateFields) {
     com.google.firestore.v1.Target.QueryTarget encodedQueryTarget =
         serializer.encodeQueryTarget(query.toTarget());
 
     StructuredAggregationQuery.Builder structuredAggregationQuery =
         StructuredAggregationQuery.newBuilder();
     structuredAggregationQuery.setStructuredQuery(encodedQueryTarget.getStructuredQuery());
 
-    StructuredAggregationQuery.Aggregation.Builder aggregation =
-        StructuredAggregationQuery.Aggregation.newBuilder();
-    aggregation.setCount(StructuredAggregationQuery.Aggregation.Count.getDefaultInstance());
-    aggregation.setAlias("count_alias");
-    structuredAggregationQuery.addAggregations(aggregation);
+    // We use a Set here to automatically remove duplicates.
+    Set<StructuredAggregationQuery.Aggregation> aggregations = new HashSet<>();
+    for (AggregateField aggregateField : aggregateFields) {
+      StructuredAggregationQuery.Aggregation.Builder aggregation =
+          StructuredAggregationQuery.Aggregation.newBuilder();
+      StructuredQuery.FieldReference fieldPath =
+          StructuredQuery.FieldReference.newBuilder()
+              .setFieldPath(aggregateField.getFieldPath())
+              .build();
+
+      if (aggregateField instanceof AggregateField.CountAggregateField) {
+        aggregation.setCount(StructuredAggregationQuery.Aggregation.Count.getDefaultInstance());
+      } else if (aggregateField instanceof AggregateField.SumAggregateField) {
+        aggregation.setSum(
+            StructuredAggregationQuery.Aggregation.Sum.newBuilder().setField(fieldPath).build());
+      } else if (aggregateField instanceof AggregateField.AverageAggregateField) {
+        aggregation.setAvg(
+            StructuredAggregationQuery.Aggregation.Avg.newBuilder().setField(fieldPath).build());
+      } else {
+        throw new RuntimeException("Unsupported aggregation");
+      }
+
+      aggregation.setAlias(aggregateField.getAlias());
+      aggregations.add(aggregation.build());
+    }
+    structuredAggregationQuery.addAllAggregations(aggregations);
```

Comment:
Wondering if all of this logic for creating the StructuredAggregationQuery should be in RemoteSerializer.java?

Reply:
Great point!